File: src/main/java/de/rubixdev/enchantedshulkers/config/WorldConfig.java
package de.rubixdev.enchantedshulkers.config;

import static java.nio.file.StandardOpenOption.*;

import com.google.common.collect.Lists;
import com.moandjiezana.toml.Toml;
import com.moandjiezana.toml.TomlWriter;
import de.rubixdev.enchantedshulkers.Mod;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.lang.reflect.Field;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.stream.Stream;

import net.fabricmc.loader.api.FabricLoader;
import net.fabricmc.loader.api.Version;
import net.fabricmc.loader.api.VersionParsingException;
import net.minecraft.resource.ResourcePackManager;
import net.minecraft.resource.ResourcePackProfile;
import net.minecraft.server.MinecraftServer;
import net.minecraft.text.Text;
import net.minecraft.util.WorldSavePath;

public class WorldConfig {
    private static MinecraftServer server;
    private static Inner inner = new Inner();

    public static void attachServer(MinecraftServer server) {
        WorldConfig.server = server;
        WorldConfig.read();
        updateResources();
    }

    public static void detachServer() {
        WorldConfig.server = null;
        WorldConfig.inner = new Inner();
    }

    public static Stream<String> getOptions() {
        return Arrays.stream(Inner.class.getDeclaredFields()).map(Field::getName);
    }

    public static Object getOption(String option) {
        Field optionField = getField(option);
        try {
            return optionField.get(inner);
        } catch (IllegalAccessException e) {
            // shouldn't fail after `setAccessible(true)` call
            throw new RuntimeException(e);
        }
    }

    public static Object getOptionDefault(String option) {
        Field optionField = getField(option);
        try {
            return optionField.get(new Inner());
        } catch (IllegalAccessException e) {
            // shouldn't fail after `setAccessible(true)` call
            throw new RuntimeException(e);
        }
    }

    public static boolean getBooleanValue(String option) {
        Object value = getOption(option);
        if (value instanceof Boolean bool) return bool;
        if (value instanceof Number number) return number.doubleValue() > 0;
        return false;
    }

    public static void setOption(String option, String value) throws InvalidOptionValueException {
        Field optionField = getField(option);
        try {
            if (optionField.getType() == boolean.class || Boolean.class.isAssignableFrom(optionField.getType())) {
                if (value.equals("true")) optionField.setBoolean(inner, true);
                else if (value.equals("false")) optionField.setBoolean(inner, false);
                else throw new InvalidOptionValueException(Text.translatable("commands." + Mod.MOD_ID + ".invalid_bool", value));
            } else if (optionField.getType() == int.class || Integer.class.isAssignableFrom(optionField.getType())) {
                optionField.setInt(inner, getIntValue(optionField, value));
            } else {
                throw new IllegalStateException("Option '" + option + "' has an unsupported type");
            }
            updateResources();
        } catch (IllegalAccessException e) {
            // shouldn't fail after `setAccessible(true)`
            throw new RuntimeException(e);
        }
        write();
    }

    private static int getIntValue(Field optionField, String value) throws InvalidOptionValueException {
        int intValue;
        try {
            intValue = Integer.parseInt(value);
        } catch (NumberFormatException e) {
            throw new InvalidOptionValueException(Text.translatable("commands." + Mod.MOD_ID + ".invalid_int", value));
        }
        IntOption intOption = optionField.getAnnotation(IntOption.class);
        if (intOption != null && (intValue < intOption.min() || intValue > intOption.max())) {
            throw new InvalidOptionValueException(Text.translatable("commands." + Mod.MOD_ID + ".invalid_int_range", intOption.min(), intOption.max(), intValue));
        }
        return intValue;
    }

    public static Field getField(String name) {
        try {
            Field field =  Inner.class.getDeclaredField(name);
            field.setAccessible(true);
            return field;
        } catch (NoSuchFieldException e) {
            throw new RuntimeException(e);
        }
    }

    private static final String ENCHANTED_ENDER_CHEST_ID;

    static {
        // Fabric API 0.95.4 and later use a different resource pack naming scheme
        try {
            ENCHANTED_ENDER_CHEST_ID = FabricLoader.getInstance().getModContainer("fabric-api").orElseThrow(RuntimeException::new).getMetadata()
                    .getVersion().compareTo(Version.parse("0.95.4")) >= 0
                ? "enchantedshulkers:enchanted_ender_chest_resourcepacks" + File.separator + "enchanted_ender_chest"
                : "enchantedshulkers:enchanted_ender_chest";
        } catch (VersionParsingException e) {
            throw new RuntimeException(e);
        }
    }

    private static void updateResources() {
        ResourcePackManager manager = server.getDataPackManager();
        boolean isEnderPackLoaded = manager.getEnabledNames().contains(ENCHANTED_ENDER_CHEST_ID);
        if (isEnderPackLoaded == inner.enchantableEnderChest) return;


        ArrayList<ResourcePackProfile> loadedPacks =
                Lists.newArrayList(server.getDataPackManager().getEnabledProfiles());
        ResourcePackProfile enderPackProfile = manager.getProfile(ENCHANTED_ENDER_CHEST_ID);
        assert enderPackProfile != null;
        if (inner.enchantableEnderChest) {
            enderPackProfile.getInitialPosition().insert(loadedPacks, enderPackProfile, profile -> profile, false);
        } else {
            loadedPacks.remove(enderPackProfile);
        }
        server.reloadResources(
                        loadedPacks.stream().map(ResourcePackProfile::getName).toList())
                .exceptionally(throwable -> {
                    Mod.LOGGER.warn("Failed to execute reload", throwable);
                    return null;
                });
    }

    private static Path getConfigPath() {
        return server.getSavePath(WorldSavePath.ROOT).resolve(Mod.MOD_ID + ".toml");
    }

    private static void read() {
        try (BufferedReader reader = Files.newBufferedReader(getConfigPath())) {
            inner = new Toml().read(reader).to(Inner.class);
            Mod.LOGGER.info("Loaded settings from " + Mod.MOD_ID + ".toml");
        } catch (Throwable e) {
            Mod.LOGGER.warn("Could not read config, using default settings: " + e);
        }
    }

    private static void write() {
        try (BufferedWriter writer = Files.newBufferedWriter(getConfigPath(), WRITE, TRUNCATE_EXISTING, CREATE)) {
            new TomlWriter().write(inner, writer);
        } catch (Throwable e) {
            Mod.LOGGER.error("Could not write config: " + e);
        }
    }

    public static boolean refillOffhand() {
        return inner.refillOffhand;
    }

    public static boolean refillNonStackables() {
        return inner.refillNonStackables;
    }

    public static boolean coloredNames() {
        return inner.coloredNames;
    }

    public static boolean creativeSiphon() {
        return inner.creativeSiphon;
    }

    public static boolean creativeRefill() {
        return inner.creativeRefill;
    }

    public static boolean creativeVacuum() {
        return inner.creativeVacuum;
    }

    public static boolean creativeVoid() {
        return inner.creativeVoid;
    }

    public static boolean generateRefill() {
        return inner.generateRefill;
    }

    public static boolean generateSiphon() {
        return inner.generateSiphon;
    }

    public static boolean generateVacuum() {
        return inner.generateVacuum;
    }

    public static boolean generateVoid() {
        return inner.generateVoid;
    }

    public static boolean generateAugment() {
        return inner.generateAugment;
    }

    public static int nestedContainers() {
        return inner.nestedContainers;
    }

    public static boolean strongerSiphon() {
        return inner.strongerSiphon;
    }

    public static boolean weakerVacuum() {
        return inner.weakerVacuum;
    }

    public static int maxAugmentLevel() {
        return inner.maxAugmentLevel;
    }

    private static class Inner {
        boolean refillOffhand = true;
        boolean refillNonStackables = false;
        boolean enchantableEnderChest = false;
        boolean coloredNames = false;
        boolean creativeSiphon = false;
        boolean creativeRefill = false;
        boolean creativeVacuum = false;
        boolean creativeVoid = false;
        boolean generateRefill = true;
        boolean generateSiphon = true;
        boolean generateVacuum = false;
        boolean generateVoid = false;
        boolean generateAugment = false;
        @IntOption(min = 0, max = Short.MAX_VALUE, suggestions = {"0", "1", "127", "255"})
        int nestedContainers = 255;
        boolean strongerSiphon = false;
        boolean weakerVacuum = false;
        @IntOption(min = 1, max = 10, suggestions = {"1", "2", "3", "6", "9", "10"})
        int maxAugmentLevel = 3;
    }
}

File: src/main/kotlin/de/rubixdev/enchantedshulkers/config/WorldConfig.kt
package de.rubixdev.enchantedshulkers.config

import com.google.gson.Gson
import com.moandjiezana.toml.Toml
import com.moandjiezana.toml.TomlWriter
import de.rubixdev.enchantedshulkers.Mod
import de.rubixdev.enchantedshulkers.Mod.MOD_ID
import de.rubixdev.enchantedshulkers.Utils.clientModVersion
import java.io.File
import java.io.IOException
import java.util.function.Function
import kotlin.reflect.KMutableProperty1
import kotlin.reflect.full.declaredMemberProperties
import kotlin.reflect.full.findAnnotation
import kotlin.reflect.full.starProjectedType
import kotlin.reflect.jvm.isAccessible
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking
import net.fabricmc.loader.api.FabricLoader
import net.fabricmc.loader.api.Version
import net.minecraft.nbt.NbtByte
import net.minecraft.nbt.NbtCompound
import net.minecraft.nbt.NbtInt
import net.minecraft.nbt.NbtString
import net.minecraft.server.MinecraftServer
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.text.Text
import net.minecraft.util.WorldSavePath

object WorldConfig {
    private var server: MinecraftServer? = null
    private var inner = Inner()
    private val GSON = Gson()

    fun attachServer(server: MinecraftServer) {
        this.server = server
        read()
        updateResources()
    }

    fun detachServer() {
        server = null
        inner = Inner()
    }

    val options: List<String> get() = Inner::class.declaredMemberProperties.map { it.name }

    fun getOption(option: String): Any? {
        val field = getField(option)
        return field.get(inner)
    }

    fun getOptionDefault(option: String): Any? {
        val field = getField(option)
        return field.get(Inner())
    }

    fun getBooleanValue(option: String): Boolean {
        return when (val value = getOption(option)) {
            is Boolean -> value
            is Number -> value.toDouble() > 0
            else -> false
        }
    }

    @Suppress("UNCHECKED_CAST")
    fun setOption(option: String, value: String) {
        val field = getField(option)
        when (field.returnType) {
            Boolean::class.starProjectedType -> {
                val boolField = field as KMutableProperty1<Inner, Boolean>
                when (value) {
                    "true" -> boolField.set(inner, true)
                    "false" -> boolField.set(inner, false)
                    else -> throw InvalidOptionValueException(
                        Text.translatable(
                            "commands.$MOD_ID.invalid_bool",
                            value,
                        ),
                    )
                }
            }

            Int::class.starProjectedType -> {
                val intField = field as KMutableProperty1<Inner, Int>
                val intValue = value.toIntOrNull()
                    ?: throw InvalidOptionValueException(Text.translatable("commands.$MOD_ID.invalid_int", value))
                val intOption = field.findAnnotation<IntOption>()
                if (intOption != null && (intValue < intOption.min || intValue > intOption.max)) {
                    throw InvalidOptionValueException(
                        Text.translatable(
                            "commands.$MOD_ID.invalid_int_range",
                            intOption.min,
                            intOption.max,
                            intValue,
                        ),
                    )
                }
                intField.set(inner, intValue)
            }

            else -> throw IllegalStateException("Option '$option' has an unsupported type: ${field.returnType}")
        }
        updateResources()
        write()
    }

    fun suggestions(option: String): Array<String> {
        val field = getField(option)
        if (field.returnType == Boolean::class.starProjectedType) return arrayOf("true", "false")
        val intOption = field.findAnnotation<IntOption>()
        if (intOption != null) return intOption.suggestions
        return arrayOf(getOptionDefault(option).toString())
    }

    private fun getField(name: String) =
        Inner::class.declaredMemberProperties.first { it.name == name }.apply { isAccessible = true }

    val OPTIONAL_PACKS = mapOf(
        "enchanted_ender_chest" to { inner.enchantableEnderChest },
        "enchanted_ender_pouch" to { inner.enchantableEnderChest && FabricLoader.getInstance().isModLoaded("things") },
        "augmented_chests_and_barrels" to { inner.augmentableChestsAndBarrels },
    )

    private val NEW_FABRIC_API = FabricLoader.getInstance().getModContainer("fabric-api")
        .orElseThrow(::RuntimeException).metadata.version >= Version.parse("0.95.4")
    private fun packId(name: String) = "enchantedshulkers:${if (NEW_FABRIC_API) "${name}_resourcepacks${File.separator}$name" else name}"

    private fun updateResources() {
        val manager = server?.dataPackManager ?: return
        val loadedPacks = manager.enabledProfiles.toMutableList()

        for ((pack, predicate) in OPTIONAL_PACKS) {
            val id = packId(pack)
            val isLoaded = manager.enabledNames.contains(id)
            val shouldBeLoaded = predicate()
            if (isLoaded == shouldBeLoaded) continue

            val packProfile = manager.getProfile(id)!!
            if (shouldBeLoaded) {
                packProfile.initialPosition.insert(loadedPacks, packProfile, Function.identity(), false)
            } else {
                loadedPacks.remove(packProfile)
            }
        }

        server!!.reloadResources(loadedPacks.map { it.name }).exceptionally {
            Mod.LOGGER.warn("Failed to execute reload", it)
            null
        }
    }

    private fun getConfigPath() = server!!.getSavePath(WorldSavePath.ROOT).resolve("$MOD_ID.toml")

    private fun read() {
        try {
            val toml = Toml().read(getConfigPath().toFile()).toMap()
            if (toml["generateBooks"] == false) {
                if ("generateSiphon" !in toml) toml["generateSiphon"] = false
                if ("generateRefill" !in toml) toml["generateRefill"] = false
            }
            toml.remove("generateBooks")
            if (toml["nestedContainers"] == true) toml["nestedContainers"] = 255
            if (toml["nestedContainers"] == false) toml["nestedContainers"] = 0
            inner = GSON.fromJson(GSON.toJsonTree(toml), Inner::class.java)
            Mod.LOGGER.info("Loaded settings from $MOD_ID.toml")
        } catch (e: Throwable) {
            Mod.LOGGER.warn("Could not read config, using default settings: $e")
        }
    }

    private fun write() {
        if (server == null) return
        try {
            TomlWriter().write(inner, getConfigPath().toFile())
            Mod.LOGGER.info("Sending updated config to clients")
        } catch (e: IOException) {
            Mod.LOGGER.error("Could not write config: $e")
        }
        server!!.playerManager.playerList.forEach(::sendConfigToClient)
    }

    fun sendConfigToClient(player: ServerPlayerEntity) {
        if (player.clientModVersion() > 0) {
            if (player.server.isHost(player.gameProfile)) {
                Mod.LOGGER.info("Not sending config to integrated server host ${player.nameForScoreboard}")
                return
            }
            Mod.LOGGER.info("Sending world config to ${player.nameForScoreboard}")
            val config = NbtCompound()
            options.forEach {
                val value = when (val value = getOption(it)) {
                    is Boolean -> NbtByte.of(value)
                    is Int -> NbtInt.of(value)
                    else -> NbtString.of(value.toString())
                }
                config.put(it, value)
            }
            ServerPlayNetworking.send(player, Mod.CONFIG_SYNC_PACKET_ID, PacketByteBufs.create().writeNbt(config))
        } else {
            Mod.LOGGER.info("Not sending config to ${player.nameForScoreboard}")
        }
    }

    @JvmStatic
    @get:JvmName("refillOffhand")
    val refillOffhand get() = inner.refillOffhand
    @JvmStatic
    @get:JvmName("refillNonStackables")
    val refillNonStackables get() = inner.refillNonStackables
    @JvmStatic
    @get:JvmName("refillProjectiles")
    val refillProjectiles get() = inner.refillProjectiles
    @JvmStatic
    @get:JvmName("coloredNames")
    val coloredNames get() = inner.coloredNames
    @JvmStatic
    @get:JvmName("creativeSiphon")
    val creativeSiphon get() = inner.creativeSiphon
    @JvmStatic
    @get:JvmName("creativeRefill")
    val creativeRefill get() = inner.creativeRefill
    @JvmStatic
    @get:JvmName("creativeVacuum")
    val creativeVacuum get() = inner.creativeVacuum
    @JvmStatic
    @get:JvmName("creativeVoid")
    val creativeVoid get() = inner.creativeVoid
    @JvmStatic
    @get:JvmName("generateRefill")
    val generateRefill get() = inner.generateRefill
    @JvmStatic
    @get:JvmName("generateSiphon")
    val generateSiphon get() = inner.generateSiphon
    @JvmStatic
    @get:JvmName("generateVacuum")
    val generateVacuum get() = inner.generateVacuum
    @JvmStatic
    @get:JvmName("generateVoid")
    val generateVoid get() = inner.generateVoid
    @JvmStatic
    @get:JvmName("generateAugment")
    val generateAugment get() = inner.generateAugment
    @JvmStatic
    @get:JvmName("nestedContainers")
    val nestedContainers get() = inner.nestedContainers
    @JvmStatic
    @get:JvmName("strongerSiphon")
    val strongerSiphon get() = inner.strongerSiphon
    @JvmStatic
    @get:JvmName("weakerVacuum")
    val weakerVacuum get() = inner.weakerVacuum
    @JvmStatic
    @get:JvmName("maxAugmentLevel")
    val maxAugmentLevel get() = inner.maxAugmentLevel
    @JvmStatic
    @get:JvmName("augmentableChestsAndBarrels")
    val augmentableChestsAndBarrels get() = inner.augmentableChestsAndBarrels

    private class Inner {
        var refillOffhand = true
        var refillNonStackables = false
        var refillProjectiles = true
        var enchantableEnderChest = false
        var coloredNames = false
        var creativeSiphon = false
        var creativeRefill = false
        var creativeVacuum = false
        var creativeVoid = false
        var generateRefill = true
        var generateSiphon = true
        var generateVacuum = false
        var generateVoid = false
        var generateAugment = false
        @IntOption(min = 0, max = Short.MAX_VALUE.toInt(), suggestions = ["0", "1", "127", "255"])
        var nestedContainers = 255
        var strongerSiphon = false
        var weakerVacuum = false
        @IntOption(min = 1, max = 10, suggestions = ["1", "2", "3", "6", "9", "10"])
        var maxAugmentLevel = 3
        var augmentableChestsAndBarrels = false
    }
}
